packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.kt

```diff
@@ -141,6 +141,15 @@ class SpeechModule(
               emitter.emit("Exponent.speakingStarted", idToMap(utteranceId))
             }
 
+            override fun onRangeStart(utteranceId: String, start: Int, end: Int, frame: Int) {
+              var map = Bundle().apply {
+                putString("id", utteranceId)
+                putInt("characterLocation", frame)
+                putInt("characterLength", start + end)
```

Comment:
Are these positions correct?
Shouldn't location be determined by `start` and length by `end - start`?

According to Android docs, the `frame` means audio frame, not the character position in string:
> `frame` - int: The position in frames in the audio of the request where this range is spoken.

Reply:
hm, you are right. I did smoke test but I think I know what happened. I smoke tested before I got the highlighting test implemented and didn't test again. I will fix.

Reply:
I see that `expo-speech` has `onBoundary` undocumented. Looking at [web MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisUtterance/boundary_event), it technically represents the same, so this might be a good idea, but it requires some more work with unifying the API and documenting it:
- Text position: iOS: ✅ , android: ✅ web: ❌ 
- Text length: iOS: ✅ , android: ✅ web: ❌ 
- Sound timestamp: iOS: ❌  , android: ✅ web: ✅

I personally have no strong opinion on this

Reply:
yes, I have been looking into those.


> Text position: iOS: ✅ , android: ✅ web: ❌

This does already exist as [`charIndex`](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisEvent/charIndex).

> Text length: iOS: ✅ , android: ✅ web: ❌

Length does not exist in the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisUtterance/boundary_event) but [`charLength`](https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L1719) does exist in this [type](https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L1717-L1723) that my IDE is able to find when I "go to definition" (a RN shim?).

> Sound timestamp: iOS: ❌ , android: ✅ web: ✅

For Android, are you referring to the [`frame`](https://developer.android.com/reference/android/speech/tts/UtteranceProgressListener#onRangeStart(java.lang.String,%20int,%20int,%20int)) parameter?

Here's what I'm considering for now:

```js
// Speech.js
setSpeakingListener('Exponent.speakingWillSayNextString', ({ id, charIndex, charLength, name }) => {
    const options = _CALLBACKS[id];
    if (options && options.onBoundary) {
        options.onBoundary({
            charIndex,
            charLength,
            name,
        });
    }
});
```

```kt
override fun onRangeStart(utteranceId: String, start: Int, end: Int, frame: Int) {
  var map = Bundle().apply {
    putString("id", utteranceId)
    putInt("charIndex", start)
    putInt("charLength", end - start)
    putInt("name", "onRangeStart")
  }
  emitter.emit("Exponent.speakingWillSayNextString", map)
}
```

```objective-c
- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer
  willSpeakRangeOfSpeechString:(NSRange)characterRange utterance:(nonnull AVSpeechUtterance *)utterance
{
  id<EXEventEmitterService> emitter = [_moduleRegistry getModuleImplementingProtocol:@protocol(EXEventEmitterService)];
  if (emitter != nil) {
    [emitter sendEventWithName:@"Exponent.speakingWillSayNextString" body:@{ @"id": ((EXSpeechUtteranceWithId *) utterance).utteranceId, @"charIndex": @(characterRange.location), @"charLength": @(characterRange.length), @"name": "willSpeakRangeOfSpeechString" }];
}
```

I'm on the fence about sending [`name`](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisEvent/name) and can remove if the semantics don't match for the native events, respectively.

Thoughts?

Reply:
Here's a [proposal](https://github.com/paper-co/expo/pull/3). I decided against using `name` or worrying about the sound timestamp parameters for now. I added caveats to the documentation similar to something I found [here](https://github.com/expo/expo/blob/main/packages/expo-device/src/Device.ts#L264-L280).